packages/core/src/common/message-rpc/websocket-client-channel.ts

```diff
@@ -0,0 +1,229 @@
+ * An attempt at a channel implementation over a websocket with fallback to http.
+ */
+
+export interface WebSocketOptions {
+    /**
+     * True by default.
+     */
+    reconnecting?: boolean;
+}
+
+export const HttpFallbackOptions = Symbol('HttpFallbackOptions');
+
+export interface HttpFallbackOptions {
+    /** Determines whether Theia is allowed to use the http fallback. True by default. */
+    allowed: boolean;
+    /** Number of failed websocket connection attempts before the fallback is triggered. 2 by default. */
+    maxAttempts: number;
+    /** The maximum duration (in ms) after which the http request should timeout. 5000 by default. */
+    pollingTimeout: number;
+    /** The timeout duration (in ms) after a request was answered with an error code. 5000 by default. */
+    errorTimeout: number;
+    /** The minimum timeout duration (in ms) between two http requests. 0 by default. */
+    requestTimeout: number;
+}
+
+export const DEFAULT_HTTP_FALLBACK_OPTIONS: HttpFallbackOptions = {
+    allowed: true,
+    maxAttempts: 2,
+    errorTimeout: 5000,
+    pollingTimeout: 5000,
+    requestTimeout: 0
+};
+
+export class WebSocketClientChannel implements Channel {
+
+    protected readonly readyDeferred: Deferred<void> = new Deferred();
+
+    protected readonly onCloseEmitter: Emitter<void> = new Emitter();
+    get onClose(): Event<void> {
+        return this.onCloseEmitter.event;
+    }
+
+    protected readonly onMessageEmitter: Emitter<ReadBufferConstructor> = new Emitter();
+    get onMessage(): Event<ReadBufferConstructor> {
+        return this.onMessageEmitter.event;
+    }
+
+    protected readonly onErrorEmitter: Emitter<any> = new Emitter();
+    get onError(): Event<any> {
+        return this.onErrorEmitter.event;
+    }
+
+    protected readonly socket: ReconnectingWebSocket;
+    protected useHttpFallback = false;
+    protected websocketErrorCounter = 0;
+    protected httpFallbackId = uuid();
+    protected httpFallbackDisconnected = true;
+
+    constructor(readonly id: string, protected readonly httpFallbackOptions: HttpFallbackOptions | undefined) {
+        const url = this.createWebSocketUrl('/services');
+        const socket = this.createWebSocket(url);
+        socket.onerror = event => this.handleSocketError(event);
+        socket.onopen = () => {
+            this.fireSocketDidOpen();
+        };
+        socket.onclose = ({ code, reason }) => {
+            this.onCloseEmitter.fire();
+        };
+        socket.onmessage = ({ data }) => {
+            this.onMessageEmitter.fire(() => new ArrayBufferReadBuffer(data));
+        };
+        this.socket = socket;
+        window.addEventListener('offline', () => this.tryReconnect());
+        window.addEventListener('online', () => this.tryReconnect());
+    }
+
+    getWriteBuffer(): WriteBuffer {
+        const result = new ArrayBufferWriteBuffer();
+        const httpUrl = this.createHttpWebSocketUrl('/services');
+        if (this.useHttpFallback) {
```

Comment:
The whole HTTP fallback shouldn't be required anymore?

We're using `socket.io` which should handle this natively.

Reply:
Good catch. Yes this file was a remainder from the initial message-rpc PoC code drop. It has been removed as part of the update that I pushed.